xclim/core/units.py

```diff
@@ -302,6 +314,110 @@ def convert_units_to(
     raise NotImplementedError(f"Source of type `{type(source)}` is not supported.")
 
 
+FREQ_UNITS = {
+    "N": "nanosecond",
+    "L": "millisecond",
+    "S": "second",
+    "T": "minute",
+    "H": "hour",
+    "D": "day",
+    "W": "week",
+    "M": "month",
+    "A": "year",
+}
+
+
+def infer_sampling_units(da: xr.DataArray, deffreq: str = "D") -> Tuple[int, str]:
+    """Infer a multiplicator and the units corresponding to one sampling period.
+
+    If `xr.infer_freq` fails, returns defval
+    """
+    freq = xr.infer_freq(da.time)
+    if freq is None:
+        freq = deffreq
+
+    multi, base, _ = parse_offset(freq)
+    try:
+        return int(multi or "1"), FREQ_UNITS[base]
+    except KeyError:
+        raise ValueError(f"Sampling frequency {freq} has no corresponding units.")
+
+
+def to_agg_units(
+    out: xr.DataArray, orig: xr.DataArray, op: str, dim="time"
+) -> xr.DataArray:
+    """Set and convert units of an array after an aggregation operation along the sampling dimension (time).
+
+    Parameters
+    ----------
+    out : xr.DataArray
+      The output array of the aggregation operation, no units operation done yet.
+    orig : xr.DataArray
+      The original array before the aggregation operation,
+      used to infer the sampling units and get the variable units.
+    op : {'count', 'prod', 'delta_prod'}
+      The type of aggregation operation performed. The special "delta_*" ops are used
+      with temperature units needing conversion to their "delta" counterparts (e.g. degree days)
+    dim : str
+      The time dimension along which the aggregation was performed.
+    """
+    m, freq_u_raw = infer_sampling_units(orig[dim])
+    freq_u = str2pint(freq_u_raw)
+    orig_u = str2pint(orig.units)
+
+    out = out * m
+    if op == "count":
+        out.attrs["units"] = freq_u_raw
+    elif op == "prod":
+        out.attrs["units"] = pint2cfunits(orig_u * freq_u)
+    elif op == "delta_prod":
+        out.attrs["units"] = pint2cfunits((orig_u - orig_u) * freq_u)
+    else:
+        raise ValueError(f"Aggregation op {op} not in [count, prod, delta_prod].")
+    return out
+
+
+def rate2amount(
```

Comment:
I see no explicit test targeting this function. I understand it's tested through indices, but I suggest we check corner cases with a dedicated test.

Reply:
Could you also include a conceptual example in the docstring to clarify what this does in practical terms.